python/whylogs/core/metrics/multimetric.py

```diff
@@ -0,0 +1,170 @@
+
+    See UnicodeRangeMetric for an example.
+    """
+
+    submetrics: Dict[str, Dict[str, Metric]]
+
+    def __init__(self, submetrics: Dict[str, Dict[str, Metric]]):
+        """
+                Parameters
+                ----------
+                submetrics : Dict[str, Dict[str, Metric]]
+                    The collection of metrics that comprise the CompoundMetric.
+
+        TODO: type this
+
+                       "<submetric name>/<component name>"
+                    Submetric names should only contain alphanumeric characters,
+                    hyphens, and underscores.
+        """
+
+        if ":" in self.namespace or "/" in self.namespace:
+            raise ValueError(f"Invalid namespace {self.namespace}")
+        for sub_name in submetrics.keys():
+            if ":" in sub_name or "/" in sub_name:
+                raise ValueError(f"Invalid submetric name {sub_name}")
+
+        self.submetrics = deepcopy(submetrics)
+
+    def merge_submetrics(self: MULTI_METRIC, other: MULTI_METRIC) -> Dict[str, Metric]:
+        if self.namespace != other.namespace:
+            raise ValueError(f"Attempt to merge CompoundMetrics {self.namespace} and {other.namespace}")
+
+        submetric_names = set(self.submetrics.keys())
+        submetric_names.update(other.submetrics.keys())
+        submetrics: Dict[str, Dict[str, Metric]] = dict()
+        for submetric_name in submetric_names:
+            if submetric_name in self.submetrics and submetric_name in other.submetrics:
+                submetrics[submetric_name] = _do_submetric_merge(
+                    self.submetrics[submetric_name], other.submetrics[submetric_name]
+                )
+            elif submetric_name in self.submetrics:
+                submetrics[submetric_name] = deepcopy(self.submetrics[submetric_name])
+            else:
+                submetrics[submetric_name] = deepcopy(other.submetrics[submetric_name])
+
+        return submetrics
+
+    def merge(self: MULTI_METRIC, other: MULTI_METRIC) -> MULTI_METRIC:
+        return self.__class__(self.merge_submetrics(other))
+
+    def to_protobuf(self) -> MetricMessage:
+        msg = {}
+        for sub_name, metrics in self.submetrics.items():
+            for namespace, metric in metrics.items():
+                sub_msg = metric.to_protobuf()
+                for comp_name, comp_msg in sub_msg.metric_components.items():
+                    msg[f"{sub_name}:{namespace}/{comp_name}"] = comp_msg
+
+        return MetricMessage(metric_components=msg)
+
+    def get_component_paths(self) -> List[str]:
+        res = []
+        for sub_name, metrics in self.submetrics.items():
+            for namespace, metric in metrics.items():
+                for comp_name in metric.get_component_paths():
+                    res.append(f"{sub_name}:{namespace}/{comp_name}")
+
+        return res
+
+    def to_summary_dict(self, cfg: SummaryConfig) -> Dict[str, Any]:
```

Comment:
we should make `cfg` here optional

Reply:
I'll do a separate PR to make it optional for all metrics. It looks like it's currently required everywhere.